include/guidance/intersection_handler.hpp

```diff
@@ -123,491 +139,473 @@ class IntersectionHandler
-        const auto &lhs_data = node_data_container.GetAnnotation(
-            node_based_graph.GetEdgeData(lhs.eid).annotation_data);
-        const auto &rhs_data = node_data_container.GetAnnotation(
-            node_based_graph.GetEdgeData(rhs.eid).annotation_data);
-        const auto lhs_continues = IsContinueRoad(lhs_data);
-        const auto rhs_continues = IsContinueRoad(rhs_data);
-        const auto left_tie = std::tie(lhs.entry_allowed, lhs_continues);
-        const auto right_tie = std::tie(rhs.entry_allowed, rhs_continues);
-        return left_tie > right_tie || (left_tie == right_tie && RoadCompare(lhs, rhs));
+
+    auto const lanes_number_reduced = [&](auto const &compare_data) {
+        // Check if the lanes number is reduced going from the inbound edge to the compare road
+        return num_lanes(compare_data) > 0 &&
+               num_lanes(compare_data) + 1 < num_lanes(via_edge_data);
     };
 
-    auto best_option_it = std::min_element(begin(intersection), end(intersection), RoadCompare);
+    auto const road_has_lower_class = [&](auto const &road) {
+        // Check if a road has a strictly lower category and can be ignored:
+        auto const &compare_data = node_based_graph.GetEdgeData(road.eid);
+
+        // 1) if the road has strictly less classification than the incoming candidate roads
+        //    and has smaller number of lanes
+        if (strictlyLess(compare_data.flags.road_classification,
+                         via_edge_data.flags.road_classification) &&
+            strictlyLess(compare_data.flags.road_classification,
+                         candidate_data.flags.road_classification) &&
+            lanes_number_reduced(compare_data))
+        {
+            return true;
+        }
 
-    // min element should only return end() when vector is empty
-    BOOST_ASSERT(best_option_it != end(intersection));
+        // 2) if a link of the same category
+        if (isLinkTo(compare_data.flags.road_classification,
+                     via_edge_data.flags.road_classification) &&
+            isLinkTo(compare_data.flags.road_classification,
+                     candidate_data.flags.road_classification))
+        {
+            return true;
+        }
 
-    best_option = std::distance(begin(intersection), best_option_it);
-    best_option_deviation = angularDeviation(intersection[best_option].angle, STRAIGHT_ANGLE);
-    const auto &best_option_edge = node_based_graph.GetEdgeData(intersection[best_option].eid);
-    const auto &best_option_data =
-        node_data_container.GetAnnotation(best_option_edge.annotation_data);
+        return false;
+    };
 
-    // Unless the in way is also low priority, it is generally undesirable to
-    // indicate that a low priority road is obvious
-    if (IsLowPriority(best_option_edge) &&
-        best_option_edge.flags.road_classification != in_way_edge.flags.road_classification)
+    if (candidate_deviation < GROUP_ANGLE)
```

Comment:
Nit pick: It might make sense to split this function into two sub-functions each for handling a case of this `if` block. The function is quite huge.